base/MAPL_VerticalMethods.F90

```diff
@@ -248,24 +254,130 @@ subroutine setup_eta_to_pressure(this,regrid_handle,output_grid,rc)
           _VERIFY(status)
        end if
        deallocate(orig_surface_level)
+
+       call init_indices(_RC)
+
        _RETURN(_SUCCESS)
 
+       contains
+          ! initialize for pl3d ( not ple3d)
+          subroutine init_indices(rc)
+            integer, optional, intent(inout) :: rc
+            integer :: status
+            integer :: k, lev, km, D1, D2, levo, km_e
+            integer :: flip_sign, i,j
+            real, allocatable :: pb(:,:), pt(:,:), ple3d(:,:,:)
+            real :: pp
+
+            D1   = size(this%pl3d,1)
+            D2   = size(this%pl3d,2)
+            km   = size(this%pl3d,3)
+            levo = size(this%interp_levels)
+            flip_sign = 1
+            if( .not. this%ascending ) flip_sign = -1
+
+            ! for cell values
+            if(allocated(this%ks))  deallocate(this%ks)
+            if(allocated(this%weight)) deallocate(this%weight)
+            allocate(this%ks(D1,D2,levo),source  = -1)
+            allocate(this%weight(D1,D2,levo),source = 0.0)
+
+            do lev =1, levo
+               pp = flip_sign*this%interp_levels(lev)
+               pb = flip_sign*this%pl3d(:,:,km)
+               do k = km-1, 1,-1 ! levels of input
+                 if(all(pb<pp)) exit
+                 pt = flip_sign*this%pl3d(:,:,k)
+                 where (pp>pt .and. pp<=pb)
+                    this%ks(:,:,lev) = k
+                    this%weight(:,:,lev) = (pb-pp)/(pb-pt)
+                 end where
+                 pb = pt
+               enddo
+            enddo
+            deallocate(this%pl3d) ! not needed any more. release the memory
+
+            ! for edge values
+            if(allocated(this%ks_e))  deallocate(this%ks_e)
+            if(allocated(this%weight_e)) deallocate(this%weight_e)
```

Comment:
This is a bit unusual.   Do you have a situation where the object is initialized a 2nd time?  As opposed to a new object?

Reply:
I actually am not sure if this object is reused in different collections which may have different dimensions. This is just to be on the safe side.  The other arrays  (pl3d, ple3d) are reallocated, so I follow that logic.